clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreAwarePartitionWiseKafkaConsumerService.java

```diff
@@ -0,0 +1,196 @@
+        pubSubDeserializer,
+        time,
+        stats,
+        isKafkaConsumerOffsetCollectionEnabled,
+        metadataRepository,
+        isUnregisterMetricForDeletedStoreEnabled,
+        StoreAwarePartitionWiseKafkaConsumerService.class.toString());
+    this.storeRepository = metadataRepository;
+  }
+
+  @Override
+  protected synchronized SharedKafkaConsumer pickConsumerForPartition(
+      PubSubTopic versionTopic,
+      PubSubTopicPartition topicPartition) {
+    String resourceIdentifier = getResourceIdentifier(versionTopic);
+    int minLoad = Integer.MAX_VALUE;
+    SharedKafkaConsumer minLoadConsumer = null;
+    Map<Integer, Integer> consumerToLoadMap = new VeniceConcurrentHashMap<>();
+    for (SharedKafkaConsumer consumer: getConsumerToConsumptionTask().keySet()) {
+      int index = getConsumerToConsumptionTask().indexOf(consumer);
+      if (topicPartition.getPubSubTopic().isRealTime()
+          && alreadySubscribedRealtimeTopicPartition(consumer, topicPartition)) {
+        getLOGGER().info(
+            "Consumer id: {} has already subscribed the same real time topic-partition: {}, will assign MAX load to avoid being picked.",
+            index,
+            topicPartition);
+        continue;
+      }
+      int baseAssignmentCount = consumer.getAssignmentSize();
+      int storeSubscriptionCount =
+          getResourceIdentifierToConsumerMap().getOrDefault(resourceIdentifier, Collections.emptyMap())
+              .getOrDefault(consumer, 0);
+      int overallLoad = storeSubscriptionCount * IMPOSSIBLE_MAX_PARTITION_COUNT_PER_CONSUMER + baseAssignmentCount;
+      if (overallLoad < minLoad) {
+        minLoadConsumer = consumer;
+        minLoad = overallLoad;
+      }
+      consumerToLoadMap.put(index, overallLoad);
+    }
+    if (minLoad == Integer.MAX_VALUE) {
+      throw new IllegalStateException(
+          "Unable to find least loaded consumer entry. Current load map: " + consumerToLoadMap);
+    }
+    // Update resource identifier to consumer load map.
+    Map<PubSubConsumerAdapter, Integer> consumerMap = getResourceIdentifierToConsumerMap()
+        .computeIfAbsent(resourceIdentifier, key -> new VeniceConcurrentHashMap<>());
+    int existCount = consumerMap.getOrDefault(minLoadConsumer, 0);
+    consumerMap.put(minLoadConsumer, existCount + 1);
+
+    // Update RT topic partition consumer map.
+    if (topicPartition.getPubSubTopic().isRealTime()) {
+      getRtTopicPartitionToConsumerMap().computeIfAbsent(topicPartition, key -> new HashSet<>()).add(minLoadConsumer);
+    }
+
+    getLOGGER().info(
+        "Picked consumer id: {}, assignment size: {}, computed load: {} for topic partition: {}, version topic: {}",
+        getConsumerToConsumptionTask().indexOf(minLoadConsumer),
+        minLoadConsumer.getAssignmentSize(),
+        minLoad,
+        topicPartition,
+        versionTopic);
+    return minLoadConsumer;
+  }
+
+  @Override
+  void handleUnsubscription(
```

Comment:
Unsubscription can happen concurrently with function: `pickConsumerForPartition`, that means the logic in `pickConsumerForPartition` won't be accurate, so is it expected?
Maybe it doesn't matter that much, but I was wondering whether we have considered such behavior or not.

Reply:
yeah, they are guarded per shared consumer instance.
I guess this would be fine as the partition assignment strategy is still a best effort.